qiskit/primitives/containers/bit_array.py

```
@@ -0,0 +1,337 @@
+
+    @staticmethod
+    def from_samples(
+        samples: Iterable[str] | Iterable[int], num_bits: int | None = None
+    ) -> "BitArray":
+        """Construct a new bit array from an iterable of bitstrings, hexstrings, or integers.
+
+        All samples are assumed to be integers if the first one is. Strings are all assumed to be
+        bitstrings whenever the first string doesn't start with ``"0x"``.
+
+        Consider pairing this method with :meth:`~reshape` if your samples represent nested data.
+
+        Args:
+            samples: A list of bitstrings, a list of integers, or a list of hexstrings.
+            num_bits: The desired number of bits per sample. If unset, the biggest sample provided
+                is used to determine this value.
+
+        Returns:
+            A new bit array.
+
+        Raises:
+            ValueError: If no strings are given.
+        """
+        samples = iter(samples)
+        try:
+            first_sample = next(samples)
+        except StopIteration as ex:
+            raise ValueError("At least one sample is required.") from ex
+
+        ints = chain([first_sample], samples)
+        if isinstance(first_sample, str):
+            base = 16 if first_sample.startswith("0x") else 2
+            ints = (int(val, base=base) for val in ints)
+
+        if num_bits is None:
+            # we are forced to prematurely look at every iterand in this case
+            ints = list(ints)
+            num_bits = max(map(int.bit_length, ints))
+
+        num_bytes = _min_num_bytes(num_bits)
+        data = b"".join(val.to_bytes(num_bytes, "big") for val in ints)
+        array = np.frombuffer(data, dtype=np.uint8, count=len(data))
+        return BitArray(array.reshape(-1, num_bytes), num_bits)
+
+    def get_counts(self, loc: int | Tuple[int, ...] | None = None) -> Dict[str, int]:
+        """Return a counts dictionary.
+
+        Args:
+            loc: Which entry of this array to return a dictionary for.
```

Should loc have a default value if it will lead to errors in non-trivial cases (`shape!=()`). Another perhaps more useful implementation of this method would be to treat this as a sum over axes when computing the counts. Then `loc=None` would returned the counts for all values regardless of shape.

For shaped arrays if loc is not None you could either raise an exception if it is the incorrect lenght tuple, or do a broadcasting sum. For the later case this would be for example, if shape = (2, 2), `loc=0`, would be treated like `(slice(None), 0)`. Follow up are slices supported in `loc` to represent summation?

---

Your suggested behavour is fine with me. Note that at the moment this can be achieved with `bit_array.reshape(-1).get_counts()`, if I am not mistaken (it's been a little while).

---

Done in b25043dc8